compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -534,47 +679,24 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(implicit ctx: Context) {
       else None
   }
 
-  /** A typer for inlined code. Its purpose is:
+  /** A base trait of InlineTyper and InlineReTyper containing operations that
+   *  work the same way for both. Beyond typing or retyping, an inline typer performs
+   *  the following functions:
+   *
    *  1. Implement constant folding over inlined code
    *  2. Selectively expand ifs with constant conditions
    *  3. Inline arguments that are by-name closures
    *  4. Make sure inlined code is type-correct.
    *  5. Make sure that the tree's typing is idempotent (so that future -Ycheck passes succeed)
    */
-  private object InlineTyper extends ReTyper {
-
-    override def ensureAccessible(tpe: Type, superAccess: Boolean, pos: Position)(implicit ctx: Context): Type = {
-      tpe match {
-        case tpe: NamedType if !tpe.symbol.isAccessibleFrom(tpe.prefix, superAccess) =>
-          tpe.info match {
-            case TypeAlias(alias) => return ensureAccessible(alias, superAccess, pos)
-            case info: ConstantType if tpe.symbol.isStable => return info
-            case _ =>
-          }
-        case _ =>
-      }
-      super.ensureAccessible(tpe, superAccess, pos)
-    }
-
-    override def typedIdent(tree: untpd.Ident, pt: Type)(implicit ctx: Context) =
-      tree.asInstanceOf[tpd.Tree] match {
-        case InlineableArg(rhs) => inlining.println(i"inline arg $tree -> $rhs"); rhs
-        case _ => super.typedIdent(tree, pt)
-      }
-
-    override def typedSelect(tree: untpd.Select, pt: Type)(implicit ctx: Context): Tree = {
-      assert(tree.hasType, tree)
-      val qual1 = typed(tree.qualifier, selectionProto(tree.name, pt, this))
-      val res = untpd.cpy.Select(tree)(qual1, tree.name).withType(tree.typeOpt)
-      ensureAccessible(res.tpe, tree.qualifier.isInstanceOf[untpd.Super], tree.pos)
-      res
-    }
+  trait InlineTyping extends Typer {
 
     override def typedIf(tree: untpd.If, pt: Type)(implicit ctx: Context) = {
       val cond1 = typed(tree.cond, defn.BooleanType)
       cond1.tpe.widenTermRefExpr match {
         case ConstantType(Constant(condVal: Boolean)) =>
-          val selected = typed(if (condVal) tree.thenp else tree.elsep, pt)
+          var selected = typed(if (condVal) tree.thenp else tree.elsep, pt)
+          if (selected.isEmpty) selected = tpd.Literal(Constant(()))
```

Comment:
We don't need to type it before testing `selected.isEmpty`, I would write it this way
```scala
val selected0 = if (condVal) tree.thenp else tree.elsep
val selected = if (selected.isEmpty) tpd.Literal(Constant(())) else typed(selected0, pt)
```
Also no need for a `var`.